x-pack/plugins/lists/server/services/extension_points/extension_point_storage.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { Logger } from 'kibana/server';
+
+import { ExtensionPoint, NarrowExtensionPointToType } from './types';
+import {
+  ExtensionPointStorageClient,
+  ExtensionPointStorageClientInterface,
+} from './extension_point_storage_client';
+
+export class ExtensionPointStorage {
+  private readonly store = new Map<ExtensionPoint['type'], Set<ExtensionPoint>>();
+  private readonly registeredFrom = new Map<ExtensionPoint, string>();
+
+  constructor(private readonly logger?: Logger) {}
+
+  add(extension: ExtensionPoint): void {
+    if (!this.store.has(extension.type)) {
+      this.store.set(extension.type, new Set());
+    }
+
+    const extensionPointsForType = this.store.get(extension.type);
+
+    if (extensionPointsForType) {
+      extensionPointsForType.add(extension);
+
+      // Capture stack trace from where this extension point was registered, so that it can be used when
+      // errors occur or callbacks don't return the expected result
+      const from = new Error('REGISTERED FROM:').stack ?? 'REGISTERED FROM: unknown';
+      this.registeredFrom.set(
+        extension,
+        from.substring(from.indexOf('REGISTERED FROM:')).concat('\n    ----------------------')
+      );
+    }
+  }
+
+  clear(): void {
+    this.store.clear();
+    this.registeredFrom.clear();
+  }
+
+  getExtensionRegistrationSource(extensionPoint: ExtensionPoint): string | undefined {
+    return this.registeredFrom.get(extensionPoint);
+  }
+
+  get<T extends ExtensionPoint['type']>(
+    extensionType: T
+  ): Set<NarrowExtensionPointToType<T>> | undefined {
+    const extensionDefinitions = this.store.get(extensionType);
+
+    if (extensionDefinitions) {
+      return extensionDefinitions as Set<NarrowExtensionPointToType<T>>;
+    }
+
+    return undefined;
+  }
+
+  /**
+   * returns a client interface that does not expose the full set of methods available in the storage
+   */
+  getClient(): ExtensionPointStorageClientInterface {
+    return new ExtensionPointStorageClient(this, this.logger);
+  }
+}
+
+export type ExtensionPointStorageInterface = ExtensionPointStorage;
```

Comment:
Shouldn't ExtensionPointStorageClientInterface be a real interface with methods instead of just typed back to the class?

Reply:
I think I created this initially thinking I would need to expose the interface our of the server types so that we did not have to `export` the class, only the interface. But really, we don't need it.

I can remove it if you like or change it to a explicit interface and have the class `implements` it. What would you prefer? ( 👈  @FrankHassanabad )

Reply:
I would go ahead and use an explicit interface interface type for this and then have the class implement it and use the interface instead of the type. It looks like it would be a good fit.